### PR TITLE
Create a skeleton of a Django project

### PR DESCRIPTION
**Pull request by slsh1o**
Init a first Django app and fix all raised by linters issue.

Need this commits on master branch to proceed with tutorial.

Ask for review also: @BogdanShevchenkoo @ZVitaliy @dark-lord671

### CODE REVIEW

**Comment by slsh1o** [comment]
@webknjaz ahh, yes. As you've said build failed.
Need help with it.
Received this error on travis-ci:
```console
pylint...................................................................Failed
- hook id: pylint
- exit code: 1
Executable `pylint` not found
```

**Comment by webknjaz** [comment]
@slsh1o the current setup doesn't install pylint. `local` means that it hits the current env where `pre-commit` itself is installed so it should contain `pylint` too. I'd like to see that change in a separate PR.

**Comment by webknjaz** [comment]
> Also some additional ignorecase and args was added to .gitignore and

This is not really related to *adding a skeleton*, is it? If you want it, put it in a separate PR. Although, I don't really like to customize `gitignore.io` template because if you want to update it from upstream you may lose extra changes. Are you sure you can't generate a template that contains it in that service?

> .pre-commit-config.yaml to fix C0415, C0103 and E501.

So here's the thing: before this commit, those errors don't exist, so applying this commit does not fix them in master as you claim it does. Which makes it something that doesn't add any value but adds confusion because you've created a commit that's already clean and there's no way to actually see those errors that you claim to exist.

**Comment by webknjaz** [comment]
> some additional ignorecase and args was added

Also, this is totally non-concrete: how am I supposed to guess what you changed by reading this?

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -48,5 +48,7 @@ repos:
     -   id: pylint
         name: pylint
         entry: pylint
+        args:
+        - --good-names=urlpatterns
```
Comment (webknjaz) [inline .pre-commit-config.yaml:53]: This is a good case when to change pylint's rules: when there's something that the framework you use enforces, you should respect those principles. Well done!

Nowadays I lean towards creating a proper config file for pylint rather than using args but it's fine for now, it can be done separately.

Another thing to try is https://pypi.org/project/pylint-django/ — I believe that using it will silence a number of default pytest rules that conflict with the django way of doing things.

File: manage.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+"""Django's command-line utility for administrative tasks."""
+import os
+import sys
+
+
+def main():
+    """Invoke main function."""
+    os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'mysite.settings')
+    try:
+        # pylint: disable=C0415
```
Comment (webknjaz) [inline manage.py]: You can also use more descriptive rule identifiers rather than just numeric IDs. It's easier to understand what the rule implements w/o having to google it:
```suggestion
        # pylint: disable=import-outside-toplevel
```

Comment (webknjaz) [reply manage.py]: Still, I'm not sure why this message appears here. Maybe it could go away by just installing django rather than trying to ignore things.

File: mysite/__init__.py [hunk 3]
```diff
@@ -0,0 +1 @@
+"""Django website skeleton."""
```
Comment (webknjaz) [inline mysite/__init__.py]: Well, I think the skeleton is imprecise. While it is one, it's just a temporary state. Here it's just a package. So I'd drop that last word now.

File: mysite/asgi.py [hunk 4]
```diff
@@ -0,0 +1,16 @@
+"""
+ASGI config for mysite project.
+
+It exposes the ASGI callable as a module-level variable named ``application``.
```
Comment (webknjaz) [inline mysite/asgi.py]: I don't think that the framework documentation belongs in docstrings. The generator put it here for your convenience. It's useful for you as a getting started pointer but it's useless in the context of the concrete project as it's about implementing a website, not putting links to django docs everywhere.

File: mysite/settings.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+import os
+
+# Build paths inside the project like this: os.path.join(BASE_DIR, ...)
+BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+
+
+# Quick-start development settings - unsuitable for production
+# See https://docs.djangoproject.com/en/3.0/howto/deployment/checklist/
+
+# SECURITY WARNING: keep the secret key used in production secret!
```
Comment (webknjaz) [inline mysite/settings.py]: Make sure you understand what this implies.

File: mysite/settings.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+AUTH_PASSWORD_VALIDATORS = [
+    {
+        'NAME': (
+            'django.contrib.auth.password_validation.'
+            'UserAttributeSimilarityValidator'
+        ),
+    },
+    {
+        'NAME': (
+            'django.contrib.auth.password_validation'
```
Comment (webknjaz) [inline mysite/settings.py]: I bet you're missing a dot here.

**Review by webknjaz — CHANGES_REQUESTED** [review]
Plz apply the comments I wrote to all the similar places. I didn't copy each comment to every place where the same thing needs to be adjusted in the same manner.

**Comment by slsh1o** [comment]
Ok. Another try from scratch but on the same PR :)

File: .pre-commit-config.yaml [hunk 7]
```diff
@@ -1,6 +1,6 @@
 repos:
 -   repo: https://github.com/pre-commit/pre-commit-hooks
-    rev: v2.5.0  # Use the ref you want to point at
```
Comment (webknjaz) [inline .pre-commit-config.yaml]: This change is unrelated to the purpose of this PR and should be submitted separately.

File: .pre-commit-config.yaml [hunk 8]
```diff
@@ -36,7 +36,7 @@ repos:
     -   id: remove-crlf
     -   id: forbid-tabs
     -   id: remove-tabs
-        args: [ --whitespaces-count, '4' ]  # defaults to: 4
```
Comment (webknjaz) [inline .pre-commit-config.yaml]: This change is unrelated to the purpose of this PR and should be submitted separately.

File: .pre-commit-config.yaml [hunk 9]
```diff
@@ -48,5 +48,8 @@ repos:
     -   id: pylint
         name: pylint
         entry: pylint
+        args:
+        -   --load-plugins=pylint_django
```
Comment (webknjaz) [inline .pre-commit-config.yaml]: The spacing is weird, it should be smth like
```suggestion
        - --load-plugins=pylint_django
        - --good-names=urlpatterns
```

File: mysite/asgi.py [hunk 10]
```diff
@@ -0,0 +1,9 @@
+"""ASGI config for mysite project."""
```
Comment (webknjaz) [inline mysite/asgi.py]: It's not a config but an application init.

File: mysite/settings.py [hunk 11]
```diff
@@ -0,0 +1,112 @@
+"""Django settings for mysite project."""
+import os
+
+
+BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+
+SECRET_KEY = os.getenv('KPI_SECRET_KEY')
```
Comment (webknjaz) [inline mysite/settings.py:7]: Good point, I'll post a task about improving the deployment config even more later.

File: mysite/urls.py [hunk 12]
```diff
@@ -0,0 +1,8 @@
+"""The mysite URL Configuration."""
```
Comment (webknjaz) [inline mysite/urls.py]: I'd say it's rather a mapping. But config is also ok. Except that it's weird to see it capitalized.

File: mysite/wsgi.py [hunk 13]
```diff
@@ -0,0 +1,9 @@
+"""WSGI config for mysite project."""
```
Comment (webknjaz) [inline mysite/wsgi.py]:
```suggestion
"""WSGI application initialization for mysite project."""
```

**Review by webknjaz — CHANGES_REQUESTED** [review]
Plz move things that aren't directly related to the PR purpose to a separate PR.
Also, the gating is set up to disallow merges that don't have green CI status, just like in real projects. So when I approve this PR you'll be unable to merge if it's red because merging it should turn the project from one working state into another working state. That said, fixing the lint setup in CI by installing an extra package may be considered as a thing deserving its own PR. But adding an extra plugin that improves linting of things added here (like Django-specific rules) probably belongs here.

P.S. You have a typo in one of the commit descriptions:
```diff
-becouse
+because
```

**Comment by slsh1o** [comment]
Now should be ok but I still need #24 on master to pass Travis test here.

**Comment by webknjaz** [comment]
Make sure to close&re-open this PR after merging #24. This will re-trigger CI builds.

**Review by webknjaz — APPROVED** [review]

**Comment by webknjaz** [comment]
Plz don't make foxtrot merges
